<!DOCTYPE html>
<html>
  <head>
    
      <title>Variant calling workflow - shell script</title>
    
    <link rel="shortcut icon" type="image/x-icon" href="./favicon.ico" />
<link href="./css/bootstrap/bootstrap.css" rel="stylesheet" />
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link href="./css/bootstrap/bootstrap-responsive.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="./css/swc.css" />
<link rel="stylesheet" type="text/css" href="./css/swc-bootstrap.css" />
<meta charset="UTF-8" />
<meta http-equiv="last-modified" content="" />
<!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->

    <link rel="stylesheet" type="text/css" href="./css/lesson.css" />
  </head>
  <body>
    <div class="container">
      <div class="banner">
  <a href="http://datacarpentry.org" title="Data Carpentry">
    <img alt="Data Carpentry banner" src="./img/DC1_logo_small.png" height="50px" />
  </a>
</div>


      <div class="row-fluid">
        <div class="span10 offset1">
	  <!-- start content -->
          
          <h1>Variant calling workflow - shell script</h1>
          
          <h3>Learning Objectives:</h3>

<ul>
<li>Use a series of command line tools to perform a variant calling workflow</li>
<li>Use a For loop from the previous lesson to help automate repetitive tasks</li>
<li>Group a series of sequential commands into a script to automate a workflow</li>
</ul>

<p>To get started with this lesson, we will need to grab some data from an outside
server using <code>wget</code> on the command line.</p>

<p>Make sure you are in the dc_workshop directory first</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span><span class="nb">cd</span> ~/dc_sample_data
<span class="gp">$ </span>ls 
<span class="gp">$ </span>r_genomics  sra_metadata  untrimmed_fastq  variant_calling.tar.gz
</code></pre></div>
<p>The file &#39;variant_calling.tar.gz&#39; is what is commonly called a &quot;tarball&quot;, which is
a compressed archive similar to the .zip files we have seen before.  We can decompress
this archive using the command below.</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span>tar -zxvf variant_calling.tar.gz
</code></pre></div>
<p>This will create a directory tree that contains some input data (reference genome and fastq files)
and a shell script that details the series of commands used to run the variant calling workflow.</p>

<pre>
$ tree variant_calling

variant_calling
├── data
│   ├── ref_genome
│   │   └── ecoli_rel606.fasta
│   └── trimmed_fastq
│       ├── SRR097977.fastq
│       ├── SRR098026.fastq
│       ├── SRR098027.fastq
│       ├── SRR098028.fastq
│       ├── SRR098281.fastq
│       └── SRR098283.fastq
└── run_variant_calling.sh

3 directories, 8 files

</pre>

<p>Without getting into the details yet, the variant calling workflow will do the following steps</p>

<ol>
<li>Index the reference genome for use by bwa and samtools</li>
<li>Align reads to reference genome</li>
<li>Convert the format of the alignment to sorted BAM, with some intermediate steps.</li>
<li>Calculate the read coverage of positions in the genome</li>
<li>Detect the single nucleotide polymorphisms (SNPs)</li>
<li>Filter and report the SNP variants in VCF (variant calling format)</li>
</ol>

<p>Let&#39;s walk through the commands in the workflow</p>

<p>The first command is to change to our working directory
so the script can find all the files it expects</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span><span class="nb">cd</span> ~/dc_workshop/variant_calling
</code></pre></div>
<p>Assign the name/location of our reference genome
to a variable ($genome)</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span><span class="nv">genome</span><span class="o">=</span>data/ref_genome/ecoli_rel606.fasta
</code></pre></div>
<blockquote>
<p><strong>Tip:</strong> Within the Bash shell you can create variables at any time (as we did above, and during the &#39;for&#39; loop lesson). Assign and name and the value using the assignment operator: &#39;=&#39;. You can check the the shell knows the definition of your variable by typing: echo $variable_name. </p>
</blockquote>

<p>We need to index the reference genome for bwa and samtools. bwa
and samtools are programs that are pre-installed on our server.</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash">bwa index <span class="nv">$genome</span>
samtools faidx <span class="nv">$genome</span>
</code></pre></div>
<p>Create output paths for various intermediate and result files The -p option means mkdir will create the whole path if it does not exist (no error or message will give given if it does exist)</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span>mkdir -p results/sai
<span class="gp">$ </span>mkdir -p results/sam
<span class="gp">$ </span>mkdir -p results/bam
<span class="gp">$ </span>mkdir -p results/bcf
<span class="gp">$ </span>mkdir -p results/vcf
</code></pre></div>
<p>We will now use a loop to run the variant calling work flow of each of our fastq files, so the list of command below will be execute once for each fastq files.</p>

<p>We would start the loop like this, so the name of each fastq file will by assigned to $fq</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span><span class="k">for </span>fq <span class="k">in </span>data/trimmed_fastq/<span class="k">*</span>.fastq
<span class="gp">&gt; </span><span class="k">do</span>
<span class="gp">&gt; </span><span class="c"># etc...</span>
</code></pre></div>
<p>In the script, it is a good idea to use echo for debugging/reporting to the screen</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span><span class="nb">echo</span> <span class="s2">"working with file </span><span class="nv">$fq</span><span class="s2">"</span>
</code></pre></div>
<p>This command will extract the base name of the file
(without the path and .fastq extension) and assign it
to the $base variable</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span><span class="nv">base</span><span class="o">=</span><span class="k">$(</span>basename <span class="nv">$fq</span> .fastq<span class="k">)</span>
<span class="gp">$ </span><span class="nb">echo</span> <span class="s2">"base name is </span><span class="nv">$base</span><span class="s2">"</span>
</code></pre></div>
<p>We will assign various file names to variables both
for convenience but also to make it easier to see what 
is going on in the commands below.</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span><span class="nv">fq</span><span class="o">=</span>data/trimmed_fastq/<span class="nv">$base</span><span class="se">\.</span>fastq
<span class="gp">$ </span><span class="nv">sai</span><span class="o">=</span>results/sai/<span class="nv">$base</span><span class="se">\_</span>aligned.sai
<span class="gp">$ </span><span class="nv">sam</span><span class="o">=</span>results/sam/<span class="nv">$base</span><span class="se">\_</span>aligned.sam
<span class="gp">$ </span><span class="nv">bam</span><span class="o">=</span>results/bam/<span class="nv">$base</span><span class="se">\_</span>aligned.bam
<span class="gp">$ </span><span class="nv">sorted_bam</span><span class="o">=</span>results/bam/<span class="nv">$base</span><span class="se">\_</span>aligned_sorted.bam
<span class="gp">$ </span><span class="nv">raw_bcf</span><span class="o">=</span>results/bcf/<span class="nv">$base</span><span class="se">\_</span>raw.bcf
<span class="gp">$ </span><span class="nv">variants</span><span class="o">=</span>results/bcf/<span class="nv">$base</span><span class="se">\_</span>variants.bcf
<span class="gp">$ </span><span class="nv">final_variants</span><span class="o">=</span>results/vcf/<span class="nv">$base</span><span class="se">\_</span>final_variants.vcf    
</code></pre></div>
<p>Our data are now staged.  The series of command below will run the steps of the analytical workflow</p>

<p>Align the reads to the reference genome</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span>bwa aln <span class="nv">$genome</span> <span class="nv">$fq</span> &gt; <span class="nv">$sai</span>
</code></pre></div>
<p>Convert the output to the SAM format</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span>bwa samse <span class="nv">$genome</span> <span class="nv">$sai</span> <span class="nv">$fq</span> &gt; <span class="nv">$sam</span>
</code></pre></div>
<p>Convert the SAM file to BAM format</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span>samtools view -S -b <span class="nv">$sam</span> &gt; <span class="nv">$bam</span>
</code></pre></div>
<p>Sort the BAM file</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span>samtools sort -f <span class="nv">$bam</span> <span class="nv">$sorted_bam</span>
</code></pre></div>
<p>Index the BAM file for display purposes</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span>samtools index <span class="nv">$sorted_bam</span>
</code></pre></div>
<p>Do the first pass on variant calling by counting
read coverage</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span>samtools mpileup -g -f <span class="nv">$genome</span> <span class="nv">$sorted_bam</span> &gt; <span class="nv">$raw_bcf</span>
</code></pre></div>
<p>Do the SNP calling with bcftools</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span>bcftools view -bvcg <span class="nv">$raw_bcf</span> &gt; <span class="nv">$variants</span>
</code></pre></div>
<p>Filter the SNPs for the final output</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span>bcftools view <span class="nv">$variants</span> | /usr/share/samtools/vcfutils.pl varFilter - &gt; <span class="nv">$final_variants</span>
</code></pre></div>
<h2><strong>Exercise</strong></h2>

<p>Run the script: dcuser/dc<em>sample</em>data/variant<em>calling/run</em>variant_calling.sh</p>
<div class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="gp">$ </span>bash run_variant_calling.sh
</code></pre></div>
	  <!-- end content -->
	</div>
      </div>

      <div class="footer">
  <a class="label swc-blue-bg" href="">Facebook</a>
  <a class="label swc-blue-bg" href="">Google+</a>
  <a class="label swc-blue-bg" href="">Twitter</a>
  <a class="label swc-blue-bg" href="">GitHub</a>
  <a class="label swc-blue-bg" href="./LICENSE.html">License</a>
  <a class="bugreport label swc-blue-bg" href="mailto:?subject=bug%20in%2002-variant-calling-workflow.md">Bug Report</a>
</div>

    </div>
    <!-- Le javascript
================================================== -->
<!-- Placed at the end of the document so the pages load faster -->
<script src="./js/jquery-1.9.1.min.js"></script>
<script src="./js/bootstrap/bootstrap.min.js"></script>

  </body>
</html>
